File: app/src/main/java/io/github/romangolovanov/apps/ametro/catalog/MapCatalogSerializer.java
package io.github.romangolovanov.apps.ametro.catalog;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.ObjectReader;

import io.github.romangolovanov.apps.ametro.catalog.entities.MapInfoEntity;
import io.github.romangolovanov.apps.ametro.catalog.entities.MapInfoEntityName;
import io.github.romangolovanov.apps.ametro.catalog.entities.TransportType;
import io.github.romangolovanov.apps.ametro.catalog.entities.TransportTypeHelper;

public class MapCatalogSerializer {

    private static final ObjectReader reader;

    static {
        reader = new ObjectMapper().reader();
    }

    public static MapInfoEntity[] deserializeMapInfoArray(String jsonText) throws SerializationException {
        try {
            JsonNode json = reader.readTree(jsonText);
            if(json == null){
                return new MapInfoEntity[0];
            }
            MapInfoEntity[] maps = new MapInfoEntity[json.size()];
            for (int i = 0; i < maps.length; i++) {
                JsonNode jsonMap = json.get(i);
                var mapFile = jsonMap.get("file").asText();
                maps[i] = new MapInfoEntity(
                        jsonMap.get("city_id").asInt(),
                        jsonMap.get("file").asText(),
                        jsonMap.get("latitude").asDouble(),
                        jsonMap.get("longitude").asDouble(),
                        jsonMap.get("size").asInt(),
                        jsonMap.get("timestamp").asInt(),
                        deserializeTransports(jsonMap.get("transports")),
                        jsonMap.get("uid").asText()
                );
            }
            return maps;
        }catch (Exception ex){
            throw new SerializationException(ex);
        }
    }

    public static MapInfoEntityName[] deserializeLocalization(String jsonText) throws SerializationException {
        try {
            JsonNode json = reader.readTree(jsonText);
            if(json == null){
                return new MapInfoEntityName[0];
            }
            MapInfoEntityName[] entities = new MapInfoEntityName[json.size()];
            for (int i = 0; i < entities.length; i++) {
                JsonNode city = json.get(i);
                entities[i] = new MapInfoEntityName(
                        city.get(0).asInt(),
                        city.get(1).asText(),
                        city.get(2).asText(),
                        city.get(3).asText());
            }
            return entities;
        }catch (Exception ex){
            throw new SerializationException(ex);
        }

    }

    private static TransportType[] deserializeTransports(JsonNode transports) {
        TransportType[] types = new TransportType[transports.size()];
        for(int i=0; i<types.length;i++){
            types[i] = TransportTypeHelper.parseTransportType(transports.get(i).asText());
        }
        return types;
    }
}

File: app/src/main/java/io/github/romangolovanov/apps/ametro/catalog/MapCatalogSerializer.kt
package io.github.romangolovanov.apps.ametro.catalog

import com.fasterxml.jackson.databind.ObjectMapper
import io.github.romangolovanov.apps.ametro.catalog.entities.MapInfoEntity
import io.github.romangolovanov.apps.ametro.catalog.entities.MapInfoEntityName
import io.github.romangolovanov.apps.ametro.catalog.entities.TransportType
import io.github.romangolovanov.apps.ametro.catalog.entities.TransportTypeHelper

object MapCatalogSerializer {

    private val reader = ObjectMapper().reader()

    @Throws(SerializationException::class)
    fun deserializeMapInfoArray(jsonText: String): Array<MapInfoEntity> {
        try {
            val json = reader.readTree(jsonText) ?: return emptyArray()
            return Array(json.size()) { i ->
                val jsonMap = json.get(i)
                MapInfoEntity(
                    uid = jsonMap.get("uid").asText(),
                    cityId = jsonMap.get("city_id").asInt(),
                    types = deserializeTransports(jsonMap.get("transports")),
                    fileName = jsonMap.get("file").asText(),
                    size = jsonMap.get("size").asInt(),
                    timestamp = jsonMap.get("timestamp").asInt(),
                    latitude = jsonMap.get("latitude").asDouble(),
                    longitude = jsonMap.get("longitude").asDouble()
                )
            }
        } catch (ex: Exception) {
            throw SerializationException(ex)
        }
    }

    @Throws(SerializationException::class)
    fun deserializeLocalization(jsonText: String): Array<MapInfoEntityName> {
        try {
            val json = reader.readTree(jsonText) ?: return emptyArray()
            return Array(json.size()) { i ->
                val city = json.get(i)
                MapInfoEntityName(
                    city.get(0).asInt(),
                    city.get(1).asText(),
                    city.get(2).asText(),
                    city.get(3).asText()
                )
            }
        } catch (ex: Exception) {
            throw SerializationException(ex)
        }
    }

    private fun deserializeTransports(transports: com.fasterxml.jackson.databind.JsonNode): Array<TransportType> =
        Array(transports.size()) { i -> TransportTypeHelper.parseTransportType(transports.get(i).asText()) }
}
